internal/services/compute/virtual_machine_extension_resource.go

```diff
@@ -182,6 +186,31 @@ func resourceVirtualMachineExtensionsCreateUpdate(d *pluginsdk.ResourceData, met
 		extension.VirtualMachineExtensionProperties.ProtectedSettings = &protectedSettings
 	}
 
+	instanceView, err := vmClient.InstanceView(ctx, virtualMachineId.ResourceGroup, virtualMachineId.Name)
+	if err != nil {
+		return fmt.Errorf("retrieving InstanceView for %q: %+v", virtualMachineId, err)
+	}
+
+	locks.ByID(virtualMachineId.ID())
+	isPowerStateChanged, err := ensureVirtualMachineStarted(ctx, vmClient, *virtualMachineId, instanceView)
+	if err != nil {
+		locks.UnlockByID(virtualMachineId.ID())
+		if isPowerStateChanged {
+			if innerError := restoreVirtualMachinePowerState(ctx, vmClient, *virtualMachineId, instanceView); innerError != nil {
+				return fmt.Errorf("restoring to original power state after starting %q is failed: %+v, starting err: %+v", *virtualMachineId, innerError, err)
+			}
+		}
+		return fmt.Errorf("starting %q to create/update extension: %+v", *virtualMachineId, err)
+	}
+
+	// unlock the Virtual Machine ID immediately if power state is not changed, which indicates it is in running state at the first place, so that multiple extensions could be operated at the same time
+	// if Virtual Machine is stopped or deallocated, release the lock at the end, and multiple extensions are operated one at a time
+	if isPowerStateChanged {
+		defer locks.UnlockByID(virtualMachineId.ID())
+	} else {
+		locks.UnlockByID(virtualMachineId.ID())
```

Comment:
If unlock immediately, will state change when running the following `CreateOrUpdate` or `WaitForCompletionRef`?

Reply:
the provider won't change the state later if it unlocks immediately, as in this case all the different extensions will see the VM running and don't change its state.